### ISSUE TITLE
HAProxy Enhancements

### ISSUE THREAD

**Issue by prachidamle** [rancher/rancher#7520]
### High level Specs:

Rancher will enhance its existing HAProxy Load Balancer solution to:
•	Read SSL certs from a mounted volume, with ideally a directory structure suitable for letsencrypt dehydrated.  
•	Support a way to refresh the certs.  
•	Add an API to create a virtual host

1. Ability to load the certificates from the dedicated directory of certificates which is managed by LetsEncrypt service 

User Flow:
•	User creates a load balancer and specifies the directory of the certificates as a bind volume. The directory should be available on all hosts where the load balancer is running.
•	Any volume format that is supported by a regular service in Rancher - anonymous, bindmount, named volume - should be supported by the HAProxy LB service. A definite volume path name in the container needs to be specified for the cert volume so the HAProxy service recognizes the path.
•	Rancher HAProxy should read the certificates from the user specified directory instead of the hardcoded rancher directory
•	The HAProxy will look for the following expected names of the certs, or these may be symlinks (based on the way letsencrypt dehydrated creates things). The proxy must look for cert.pem, fullchain.pem, and privkey.pem and must follow symlinks. 
•	The following is an example directory listing:
certs/lab.crate.farm$ ls –al
.
..
cert-1482181402.csr
cert-1482181402.pem
cert.csr -> cert-1482181402.csr
cert.pem -> cert-1482181402.pem
chain-1482181402.pem
chain.pem -> chain-1482181402.pem
fullchain-1482181402.pem
fullchain.pem -> fullchain-1482181402.pem
privkey-1482181402.pem
privkey.pem -> privkey-1482181402.pem

•	Rancher HAProxy should reload the certificates only when the certificate contents are updated
•	Rancher HAProxy should perform https offload using the certs from the directory
•	Strict SNI(offloaded to the proxy), unless default certificate directory is present
•	The default directory to be used can be passed as a parameter name when LB is created

**Comment by prachidamle** [comment]
Attached Sample compose files to test host-name routing with this feature.

- docker-compose.yml:

```
version: '2'
services:
  s3:
    image: sangeetha/testnewhostrouting
    stdin_open: true
    tty: true
    labels:
      io.rancher.container.pull_image: always
      io.rancher.container.hostname_override: container_name
  lb1:
    image: prachidamle/lb-service-haproxy:latest
    volumes:
    - /home/prachi/dehyrdated/certs:/certs
    ports:
    - 8086:8086/tcp
    labels:
      io.rancher.container.agent.role: environmentAdmin
      io.rancher.container.create_agent: 'true'
      io.rancher.lb_service.certDir: /certs
  s1:
    image: sangeetha/testnewhostrouting
    stdin_open: true
    tty: true
    labels:
      io.rancher.container.pull_image: always
      io.rancher.container.hostname_override: container_name
  s2:
    image: sangeetha/testnewhostrouting
    stdin_open: true
    tty: true
    labels:
      io.rancher.container.pull_image: always
      io.rancher.container.hostname_override: container_name
```

- rancher-compose.yml

```
version: '2'
services:
  s3::
    scale: 1
    start_on_create: true
  lb1:
    scale: 1
    start_on_create: true
    lb_config:
      certs:
      - test1
      - test2
      default_cert: test-647537
      port_rules:
      - hostname: test1.com
        priority: 1
        protocol: https
        service: s1
        source_port: 8086
        target_port: 80
      - hostname: test2.com
        priority: 2
        protocol: https
        service: s2
        source_port: 8086
        target_port: 80
      - hostname: ''
        priority: 3
        protocol: https
        service: s3 
        source_port: 8086
        target_port: 80
    health_check:
      healthy_threshold: 2
      response_timeout: 2000
      port: 42
      unhealthy_threshold: 3
      interval: 2000
  s1:
    scale: 1
    start_on_create: true
  s2:
    scale: 1
    start_on_create: true
```

**Comment by sangeethah** [comment]
This feature is now supported and available with rancher-server version - v1.6.0-rc4 and rancher/lb-service-haproxy:v0.7.1.

Following are the issues that were encountered and addressed while validating this feature:

https://github.com/rancher/rancher/issues/8401
https://github.com/rancher/rancher/issues/8366
https://github.com/rancher/rancher/issues/8365
https://github.com/rancher/rancher/issues/8161
